rslib/src/collection/backup.rs

```diff
@@ -341,3 +348,65 @@ fn get_zip_file_contents(file: &mut ZipFile) -> Result<Vec<u8>> {
         Err(malformed_archive_err())
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn thinning() {
+        macro_rules! backup {
+            ($year:expr, $month:expr, $day:expr) => {
+                Backup {
+                    datetime: Local.ymd($year, $month, $day).and_hms(0, 0, 0),
+                    path: PathBuf::new(),
+                }
+            };
+            ($year:expr, $month:expr, $day:expr, $hour:expr, $min:expr, $sec:expr) => {
+                Backup {
+                    datetime: Local.ymd($year, $month, $day).and_hms($hour, $min, $sec),
+                    path: PathBuf::new(),
+                }
+            };
+        }
+
+        let today = Local.ymd(2022, 2, 22);
+        let limits = BackupLimits {
+            daily: 3,
+            weekly: 2,
+            monthly: 1,
+        };
+
+        // true => should be removed
+        let backups = [
+            // grace period
+            (backup!(2022, 2, 22), false),
+            (backup!(2022, 2, 22), false),
+            (backup!(2022, 2, 21), false),
+            // daily
+            (backup!(2022, 2, 20, 6, 0, 0), true),
+            (backup!(2022, 2, 20, 18, 0, 0), false),
+            (backup!(2022, 2, 10), false),
+            (backup!(2022, 2, 9), false),
+            // weekly
+            (backup!(2022, 2, 7), true), // Monday, week already backed up
+            (backup!(2022, 2, 6, 1, 0, 0), true),
+            (backup!(2022, 2, 6, 2, 0, 0), false),
+            (backup!(2022, 1, 6), false),
+            // monthly
+            (backup!(2022, 1, 5), true),
+            (backup!(2021, 12, 24), false),
+            (backup!(2021, 12, 1), true),
+            (backup!(2021, 11, 1), true),
```

Comment:
These macros are great for cutting down on the boilerplate :-) It might be worth adding one more test here, synthesizing backups for every day for a year in a loop, and then confirming thinning works as expected in that case too.